File: pset1/credit/credit.c


#include <cs50.h>
#include <stdio.h>
#include <string.h>



char* number_to_string(long long number);
int to_int(char d);


int main(void) {

    long long card_number;
    char* card_number_str;
    int card_len;
    int sum;

    do{

        printf("Card Number: \n");
        card_number = get_long_long();

    }while(card_number < 0);

    // Convert card number to string
    card_number_str = number_to_string(card_number);
    // Number of digits of the card
    card_len = strlen(card_number_str);
    // Initialize sum to 0
    sum = 0;

    // Multiply each second/other digit by two (2) and
    // add those products' digit together.
    // NB: Starting to the last second digit.
    for(int i=card_len-2; i >= 0; i -= 2){

        int digit = to_int(card_number_str[i]);

        // Mutiply digit by two (2)
        int prod = digit * 2;

        // Add product to sum
        int rest = prod - 10;
        if(rest >= 0){
            sum += rest;
            sum++;
        }else{
            sum += prod;
        }

    }

    // Now let’s add the sum of the digits that weren’t multiplied by 2 to the sum:
    for(int i=card_len-1; i >= 0; i -= 2){
        sum += to_int(card_number_str[i]);
    }

    // check to see if the CC number is valid and in the appropriate range
    if ( sum % 10 == 0 )
    {
        if ( (card_number >= 340000000000000 && card_number < 350000000000000) || (card_number >= 370000000000000 && card_number < 380000000000000) ){
            printf("AMEX\n");
        }
        else if ( card_number >= 5100000000000000 && card_number < 5600000000000000 ) {
            printf("MASTERCARD\n");
        }

        else if ( (card_number >= 4000000000000 && card_number < 5000000000000) || (card_number >= 4000000000000000 && card_number < 5000000000000000) ) {
            printf("VISA\n");
        }
        else {
            printf("INVALID\n");
        }
    }else{
         printf("INVALID\n");
    }


    //printf("%i", sum);
    //printf("\n");

    return 0;
}


char* number_to_string(long long number) {
    char *string = malloc (256 * sizeof (char));
    sprintf(string, "%lld", number);

    return string;
}

int to_int(char d)
{
 char str[2];

 str[0] = d;
 str[1] = '\0';
 return (int) strtol(str, NULL, 10);
}

File: pset6/credit/credit.py

# This program test if a credit card number is valid
# Author: Jeanniton Monero
# email: jeanniton.mnr@gmail.com
# website: jeanniton.me




def main():
    card_number = None
    card_number_str = None
    card_len = None
    total = None

    # Get card number
    print("Number: ", end="")
    card_number = (int)(input())

    # Get string representation of card number
    card_number_str = str(card_number)
    # Number of digit of card
    card_len = len(card_number_str)
    # Initialize `total` to zero (0)
    total = 0

    # Multiply each second/other digit by two (2) and
    # add those products' digit together.
    # NB: Starting at the last second digit
    i = card_len - 2
    while i >= 0:
        digit = int(card_number_str[i])
        # Decrement `i` to two (2) step
        i -= 2

        # Multiply digit by two (2)
        prod = digit * 2

        # Add product to `total`
        rest = prod - 10
        if rest >= 0:
            total += rest
            total += 1
        else:
            total += prod

    # Let reset the variable `i`
    i = None

    # Now let's add the `total` of the digits that weren't multiplied by 2 to the `total`:
    i = card_len - 1
    while i >= 0:
        digit = int(card_number_str[i])
        total += digit
        # Decrement `i` to two (2) step
        i -= 2

    # Let reset the variable `i`
    i = None

    # Check to see if the CC number is valid and in the appropriate range
    if total % 10 == 0:
        # Check AMEX CC
        if (card_number >= 340000000000000 and card_number < 350000000000000) or (card_number >= 370000000000000 and card_number < 380000000000000):
            print("AMEX")
        # Check MASTERCARD CC
        elif card_number >= 5100000000000000 and card_number < 5600000000000000:
            print("MASTERCARD")
        # Check VISA CC
        elif (card_number >= 4000000000000 and card_number < 5000000000000) or (card_number >= 4000000000000000 and card_number < 5000000000000000):
            print("VISA")
        # No CC match: Print INVALID
        else:
            print("INVALID")
    else:
        print("INVALID")


if __name__ == "__main__":
    main()